Plugin~/WebRTCPlugin/UnityRenderEvent.cpp

```diff
@@ -254,57 +260,76 @@ static void UNITY_INTERFACE_API OnRenderEvent(int eventID, void* data)
         return;
-    Timestamp timestamp = s_clock->CurrentTime();
-    IGraphicsDevice* device = Plugin::GraphicsDevice();
-    UnityGfxRenderer gfxRenderer = device->GetGfxRenderer();
-    void* ptr = GraphicsUtility::TextureHandleToNativeGraphicsPtr(encodeData->texture, device, gfxRenderer);
-    unity::webrtc::Size size(encodeData->width, encodeData->height);
+    }
 
-    if (s_bufferPool->bufferCount() < kLimitBufferCount)
+    for (int i = 0; i < batchData->tracksCount; i++)
     {
-        std::unique_ptr<const ScopedProfiler> profiler;
-        if (s_ProfilerMarkerFactory)
-            profiler = s_ProfilerMarkerFactory->CreateScopedProfiler(*s_MarkerEncode);
+        VideoStreamTrackData* trackData = batchData->tracks[i];
+
+        if (trackData == nullptr || trackData->texture == nullptr || trackData->source == nullptr)
+            continue;
+
+        if (trackData->action == VideoStreamTrackAction::Encode)
+        {
+            RTC_DCHECK(trackData->texture);
+            RTC_DCHECK(trackData->source);
+            RTC_DCHECK_GT(trackData->width, 0);
+            RTC_DCHECK_GT(trackData->height, 0);
+
+            UnityVideoTrackSource* source = static_cast<UnityVideoTrackSource*>(trackData->source);
+            if (!s_context->ExistsRefPtr(source))
+            {
+                trackData->source = nullptr;
+                continue;
+            }
+
+            Timestamp timestamp = s_clock->CurrentTime();
+            IGraphicsDevice* device = Plugin::GraphicsDevice();
+            UnityGfxRenderer gfxRenderer = device->GetGfxRenderer();
+            void* ptr = GraphicsUtility::TextureHandleToNativeGraphicsPtr(trackData->texture, device, gfxRenderer);
+            unity::webrtc::Size size(trackData->width, trackData->height);
 
-        auto frame = s_bufferPool->CreateFrame(ptr, size, encodeData->format, timestamp);
-        source->OnFrameCaptured(std::move(frame));
+            if (s_bufferPool->bufferCount() < kLimitBufferCount)
+            {
+                std::unique_ptr<const ScopedProfiler> profiler;
+                if (s_ProfilerMarkerFactory)
+                    profiler = s_ProfilerMarkerFactory->CreateScopedProfiler(*s_MarkerEncode);
+
+                auto frame = s_bufferPool->CreateFrame(ptr, size, trackData->format, timestamp);
+                source->OnFrameCaptured(std::move(frame));
+            }
+        }
+#if 0
+        else if (trackData->action == VideoStreamTrackAction::Decode)
+        {
+            std::unique_ptr<const ScopedProfiler> profiler;
+            if (s_ProfilerMarkerFactory)
+                profiler = s_ProfilerMarkerFactory->CreateScopedProfiler(*s_MarkerDecodeCopy);
+            UnityVideoRenderer* renderer = static_cast<UnityVideoRenderer*>(trackData->source);
+            renderer->CopyBuffer(trackData->texture, trackData->width, trackData->height, trackData->format);
+        }
+#endif
```

Comment:
Should this be #if'd out?
Looks like it won't support decode with this missing.

Reply:
Yes it's just laying out the basics for adding support in a follow-up PR :) I had this code working in an older M92 branch and currently in progress of porting it over to M107 with the new render architecture, but there's enough changes so it's better land command buffer batching changes separately.

For now, VideoStreamTrackAction::Decode is ignored and the C# code is triggering regular texture CPU uploads through the plugin interface.